pkg/generator/caches.go

```diff
@@ -251,31 +260,74 @@ func generateListenersAndRouteConfigsAndClusters(
 	}
 	listeners = append(listeners, probHTTPListener)
 
-	// Add internal listeners and routes when internal cert secret is specified.
-	if cfg.Kourier.ClusterCertSecret != "" {
-		internalTLSRouteConfig := envoy.NewRouteConfig(internalTLSRouteConfigName, clusterLocalVirtualHosts)
-		internalTLSManager := envoy.NewHTTPConnectionManager(internalTLSRouteConfig.Name, cfg.Kourier)
+	// Configure TLS listener for cluster-local traffic.
+	// If there's at least one ingress that contains the TLS field, that takes precedence.
+	// If there is not, TLS will be configured using a single cert for all the services when the certificate is configured.
+	if len(localSNIMatches) > 0 {
```

Comment:
Could we refactor part of this behind a func for both local and external names, it seems there is some repetition?

Reply:
👍 I'll take a look

Reply:
Hm they are similar but not equal, differences:
- Route config is named differently on cluster-local between the two ifs, but not on external
- Filter chains methods calls are different (`newLocalEnvoyListenerWithOneCertFilterChain`)
- One cert in external is configured with env variables, one cert in cluster-local is configured using a secret from the CM

So that extracted method would have quite some params and also a passed in function to lookup the filter chains based on different methods. Not sure if we benefit enough to refactor.

Reply:
The func is too long tbh but we can always improve later.